Re-priming non-existent options triggers db queries.

peterwilsoncc opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/59738

Forked from PR https://github.com/WordPress/wordpress-develop/pull/5548 as it's likely to go in. 

Unlike the patch I put on the on the ticket, this does get `notoptions` in the original request to the cache. As `notoptions` is in the `options` group it can be done without an additional request to the caching server. For persistent caches supporting multiple gets, this should be slightly more performant.

The initial push contained only the test changes only, the failures can be seen in the [action run](https://github.com/WordPress/wordpress-develop/actions/runs/6648557168).

felixarntz commented on src/wp-includes/option.php:
```diff
@@ -258,14 +258,24 @@ function get_option( $option, $default_value = false ) {
+function wp_prime_option_caches( $options ) {
 	$alloptions     = wp_load_alloptions();
-	$cached_options = wp_cache_get_multiple( $options, 'options' );
+	$cache_keys     = (array) $options;
+	$cache_keys[]   = 'notoptions';
+	$cached_options = wp_cache_get_multiple( $cache_keys, 'options' );
+
+	// Store notoptions, remove from cached options.
+	$notoptions = $cached_options['notoptions'];
+	unset( $cached_options['notoptions'] );
```
Part of me thinks this implementation is a bit "too clever" and would be easier to follow by separating the two concerns, similarly to how `$alloptions` are retrieved separately from the actual `$options` provided.

I realize this implementation is technically a tiny bit more efficient as it avoids two separate `wp_cache_*()` retrieval calls, but I don't think that matters for performance in reality, so I personally would favor code simplicity here. That said, not a blocker either way.

Also note that the `notoptions` cache may not be set in which case it wouldn't be an array.

Here's what I'm thinking:
```suggestion
	$cached_options = wp_cache_get_multiple( $options, 'options' );
	$notoptions     = wp_cache_get( 'notoptions', 'options' );
	if ( ! is_array( $notoptions ) ) {
		$notoptions = array();
	}
```

felixarntz reviewed (dismissed):
Thanks @peterwilsoncc, this is a great catch and the implementation looks good. Let's merge #5548 first and then we can refresh this PR and consider merge for 6.4.

I only left one point of feedback, though that's just a suggestion, not a blocker.

felixarntz commented on src/wp-includes/option.php:
```diff
@@ -258,14 +258,24 @@ function get_option( $option, $default_value = false ) {
 	foreach ( $options as $option ) {
-		if ( ( ! isset( $cached_options[ $option ] ) || ! $cached_options[ $option ] ) && ! isset( $alloptions[ $option ] ) ) {
+		if (
+			( ! isset( $cached_options[ $option ] ) || ! $cached_options[ $option ] )
```
@peterwilsoncc This is not actually part of the bug ticket you raised, but now that I see this again, I am wondering whether there's another problem here. Shouldn't this line rather be:
```suggestion
			( ! isset( $cached_options[ $option ] ) || false === $cached_options[ $option ] )
```

The reason I wonder about that is that normally, only a value of `false` is interpreted as a cache miss. I think the current logic here would for example result in a query even for an option value of `0` which could be a valid value and should be a cache hit as far as I can tell.

If that's the case, maybe we could fix it in this PR too?

felixarntz commented on src/wp-includes/option.php:
```diff
@@ -258,14 +258,24 @@ function get_option( $option, $default_value = false ) {
+	$cached_options = wp_cache_get_multiple( $cache_keys, 'options' );
+
+	// Store notoptions, remove from cached options.
+	$notoptions = $cached_options['notoptions'];
```
Note that `$notoptions` is also retrieved further down in the function. Maybe we can remove that line now if we already set it up here?

peterwilsoncc commented:
@felixarntz

* The falsey options issue occurred to me overnight too, I've pushed the change and added the test `test_wp_prime_option_caches_handles_falsey_option_value()` which tests for `''`, `'0'` and `array()`
* I've removed the second get of `notoptions` and moved to the clearer approach you suggested 

While looking at the tests, I noticed a bunch of tests for options in the `notoptions` **group** rather than using the cache key. I think this is incorrect so I will work on changing those too. Case in point

https://github.com/WordPress/wordpress-develop/blob/7c4be263cdb6e093a490b19e652f7ee707440012/tests/phpunit/tests/option/wpPrimeOptionCaches.php#L212-L216

It's currently in the new test so I'll fix that while about it.

peterwilsoncc commented on src/wp-includes/option.php:
```diff
@@ -261,11 +261,19 @@ function get_option( $option, $default_value = false ) {
 function wp_prime_option_caches( $options ) {
 	$alloptions     = wp_load_alloptions();
 	$cached_options = wp_cache_get_multiple( $options, 'options' );
```
There's no tests to ensure `$options` is an array of scalars, should there be?

Each value may need a check similar to the one at the start of the `*_options` functions too

https://github.com/WordPress/wordpress-develop/blob/6fb60e44e27e1b218ad71980f6de9bb7c5dd4418/src/wp-includes/option.php#L716-L722

Edit: There's no check to avoid querying for `notoptions` and `alloptions` either, should this call `wp_protect_special_option` or similar on the array of options to be primed.

felixarntz replied:
@peterwilsoncc I don't think either of this is necessary, here's why:
* Regarding the `is_scalar && trim` bit, I can't imagine anyone passing untrimmed option names to any of the functions. Maybe that was a real concern at some point, maybe it has always been only a theoretical concern. But since this is a new function, I don't see value in it. Otherwise we may as well include this trimming in literally _any_ core function that accepts a string identifier.
* The `wp_protect_special_option()` bit is only used in functions that write to the database so far. Given that, I don't think it's needed here (e.g. `get_option()` doesn't call it either).

Both of these are IMO theoretically nice to have, but I think there's too little value in including them to actually justify it. So I would vote no.

felixarntz commented on tests/phpunit/tests/option/wpPrimeOptionCaches.php:
```diff
@@ -126,5 +160,91 @@ public function test_wp_prime_option_caches_handles_empty_notoptions_cache() {
+	public function data_wp_prime_option_caches_handles_falsey_option_value() {
+		return array(
+			'empty string' => array( '' ),
+			'string(0)'    => array( '0' ),
```
Maybe cover a `0` integer and a `0.0` float too? While the database would store those as a string too, in the cache those would in fact be different values, so worth covering them as well.
```suggestion
			'string(0)'    => array( '0' ),
			'int(0)'       => array( 0 ),
			'float(0)'     => array( 0.0 ),
```

peterwilsoncc replied:
> Maybe cover a `0` integer and a `0.0` float too? While the database would store those as a string too, in the cache those would in fact be different values, so worth covering them as well.

I realised I was testing the wrong thing.

The fundamental of the test is: re-priming an option should not trigger an additional database query. The current bug is for falsey types, but type does not matter. In 46786030eec3a5de797b24b4a6678d3d31190e86 I've refactored the entire test to test the base problem.

felixarntz reviewed (approved):
@peterwilsoncc The updated version still looks great to me, and it's great to see we were able to address two bugs here. Thank you also for fixing some of the tests that weren't right before!

One tiny suggestion for additional test cases, but not a blocker.

peterwilsoncc committed:
All the test changes.

peterwilsoncc committed:
src changes

peterwilsoncc commented:
I've forced pushed to get a run of the failing tests only as I added some after the initial source changes.

The failing test run can be seen in [this run of the GitHub actions](https://github.com/WordPress/wordpress-develop/actions/runs/6686317573).

Here's a summary of the bugs I've found:

1. repriming non-existent options triggered a database call
2. repriming falsey options triggered a database call (props Felix)
3. the shape of the caches differed when primed via the priming function vs `get_option()` (they are cached in their serialized form in `get_option()`
4. the tests for ensuring priming by group did in fact prime could result in a false pass

I've refactored my initial tests to ensure falsey values didn't trigger a repeat database call to test that no type of value triggered a repeat database call. This better represents the heart of the issue: the type doesn't matter, the additional DB call does.

cc @hellofromtonya

peterwilsoncc committed:
More types of tests.

peterwilsoncc committed:
Typos in messags.

peterwilsoncc committed:
Test items in alloptions do not trigger DB calls.

peterwilsoncc committed:
Test priming no options doesn’t make a DB query.

peterwilsoncc committed:
Ensure cache isn’t reprimed when testing if cache primed.

peterwilsoncc committed:
Test a mix of primed and unprimed options.

peterwilsoncc committed:
Test a mix of pre-primed and unprimed options.

peterwilsoncc committed:
Flush group unavailable.

felixarntz reviewed (approved):
Thank you @peterwilsoncc for identifying and already addressing the additional issue here with the additional `maybe_unserialize()` call, which should not have been present there.

This PR looks good to commit, and I think it's reasonable to backport it to the 6.4 branch even this late. While it addresses 3 bugs of varying severity, I don't see any risk of pushing these changes this late, given those are new functions.

joemcgill reviewed (approved):
Nice catch, and the additional fixes included here seem good to me too. Thanks for adding additional unit tests to cover these changes.

peterwilsoncc committed:
Merge branch 'trunk' into fix/59738-notoptions-db-queries

peterwilsoncc commented:
Committed in [r57029](https://core.trac.wordpress.org/changeset/57029) / https://github.com/WordPress/wordpress-develop/commit/17a01ed185f38f87362eb7588908dc95ec6cbe23
